IMP migrate useful internal features

mantzouratos opened the pull request:
This PR provides a variety of useful features (one per commit). Briefly:

**Catalog**
Overwrite location of external table in catalog.

**Functions (or, shortcuts)**
* "Switch/case" column generation. Can replace chained `when`s on a column value.
* Join multiple DataFrames at once and coalesce ambiguous columns across them.

**Testing**
Extend `assertEqual` to to work better with comparisons involving rows, datatypes, dictionaries, lists and floats by:
- ignoring the order of lists and datatypes recursively (useful to compare DataFrames that might contain nested arrays of structs of ... (etc) where order does not matter),
- comparing floats within a given tolerance,
- assuming NaNs are equal,
- ignoring the nullability requirements of datatypes (since Spark can be inaccurate when inferring it),
- providing better diffs for rows and datatypes.

**Writing**
Support `redis://` and `rediss://` URLs.

**Other**
* Check whether a complex `dataType` defines a minimum set of required fields. Useful for schema assertions.
* LRU DataFrame caching. Provide a unified decorator that persists DataFrames (when it encounters them) according to user settings in addition to caching the actual python object. Useful in case a user defined transformation is called in multiple places in a project and the user wants to control whether the data is being persisted to memory / disk without having to change any code.

skliarpawlo commented on sparkly/utils.py:
```diff
@@ -93,6 +181,45 @@ def kafka_get_topics_offsets(host, topic, port=9092):
     return offsets
 
 
+class lru_cache(object):
```
how much dataframes are you generating in one spark job for this to be useful?

mantzouratos replied:
Honestly I've never used the "LRU" part of this cache. I'd imagine that when targeted towards dataframes probably you'd either keep it unlimited or very small like `maxsize=1` or `2`.

srstrickland replied:
I think it's a valid use case for the event that you are maintaining dataframes programatically / dynamically, rather than having them explicitly defined individually in code.

skliarpawlo replied:
can anyone please provide some concrete real case, just don't get how this may be useful

mantzouratos replied:
Sure.

First example could be a transformation that is used multiple times in a Spark project. More concretely, let's say you have 
```
def transformation_a(...):
    # return some dataframe
```
and let's say `transformation_a` is called twice from different parts of the code. To cache the result you could just decorate it with `@lru_cache` and call it a day instead of handling it explicitly anywhere in code.

Another example would be the following. Think that you have a `settings.yaml` where you define your data sources:
```
DATA_SOURCES:
    table_a: 
        url: 'mysql://blahblah'
        cache: True

    table_b:
        url: 'parquet:s3://blahblah'
        cache: False
```
and we have a reader that loads the dataframes from the `url` sources. Here we might want to cache the MySQL table because it's read in multiple independent transformations but we don't want to keep hitting the hot storage to get it. Now again you could just decorate with `@lru_cache` the main reader code and not have to do anything else in code.

In our private code these examples correspond to https://github.com/Tubular/tbcode/blob/master/svc-creator/loader/src/creator/loader/transformations/accounts/followers_growth.py#L193-L194 and the latter is there: https://github.com/Tubular/tbcode/blob/master/svc-creator/loader/src/creator/loader/shared/io.py#L162-L229

skliarpawlo replied:
Got it, thanks. honestly I'd prefer to drop LRU part of this. It's sounds like "too much" for the feature

mantzouratos replied:
It's not the same. The current approach can handle un-persisting the dataframes on eviction from cache. If we separate them that's not possible anymore.

skliarpawlo commented on sparkly/utils.py:
```diff
@@ -59,6 +64,89 @@ def absolute_path(file_path, *rel_path):
+        yield chunk
+
+
+def data_type_defines(t, required_fields):
```
imo 'schema_has' name was more informative and less verbose at the time

mantzouratos replied:
Yeah, in the beginning I used such a name but then this function works for arbitrary types, not just `df.schema`s, so `dataType_has` or `dataType_defines` seemed better. Then I broke `dataType` to `data_type` because we don't really use camel case anywhere

srstrickland replied:
`type_has`?

mantzouratos replied:
I prefer to keep `dataType` in there since it's really working with `dataType`s and not arbitrary python types, but I'll change it to `_has`.

mantzouratos replied:
Meh ok, I ended up changing it to `type_has`

skliarpawlo commented on sparkly/utils.py:
```diff
@@ -59,6 +64,89 @@ def absolute_path(file_path, *rel_path):
     )
 
 
+def chunks(seq, n):
```
I don't agree this should be a public function of the package

mantzouratos replied:
Well we need it in `writer.py` when we're writing to Redis. The alternative option would be to include it as a private function there, which I'm not terribly opposed to, but it kind of seems like in conflict with the existence of utilities like `absolute_path` which are equally generic.

srstrickland commented on sparkly/catalog.py:
```diff
@@ -26,6 +27,57 @@ def __init__(self, spark):
         """
         self._spark = spark
 
+    def create_external_table(self, table_name, path=None, source=None, schema=None, **options):
```
In 2.1, `createExternalTable` is deprecated in favor of `createTable`. I wonder if we should make this `create_table`, optionally with a synonym for `create_external_table`

mantzouratos replied:
I don't think we need to provide a synonym

srstrickland commented on sparkly/testing.py:
```diff
@@ -120,28 +147,122 @@ def tearDown(self):
+                    # equal otherwise
+                    obj = 'nan'
+                elif places is not None:
+                    obj = round(obj, places)
```
While I see how rounding makes this whole thing easier, I don't think it's the best we can do with fuzzy float comparison.  In general it's just not a good way to do tolerance checking, especially if your values can differ by large orders of magnitude.  A percentage-based approach would be far more resilient: "values must be within 0.001%" is a much better way to specify approximate matching because this handles differing scale in a graceful way.  I admit that most of our use cases to date have been simple calculations that we can control, but 1) this is a general-purpose tool and 2) we will probably start introducing more complex calculations which will need tolerance-based comparison rather than this simpler approach.

srstrickland commented on sparkly/writer.py:
```diff
@@ -255,6 +275,173 @@ def write_partition_to_kafka(messages):
+              host=None,
+              port=6379,
+              db=0,
+              redis_cli_init=None):
```
would prefer `redis_client_init` here, as `cli` usually means "command line client"

srstrickland commented on sparkly/writer.py:
```diff
@@ -56,6 +74,7 @@ def by_url(self, url):
             - Elastic ``elastic://``
             - MySQL ``mysql://``
             - Parquet ``parquet://``
+            - Redis ``redis://`` or ``rediss://``
```
what is `rediss://` for?

mantzouratos replied:
`redis:/` and `rediss://` URLs are parsed by redis-py to set up the connection type etc. `rediss://` creates an SSL wrapped TCP socket connection. Details: http://www.iana.org/assignments/uri-schemes/prov/rediss

skliarpawlo commented on sparkly/testing.py:
```diff
@@ -120,28 +147,132 @@ def tearDown(self):
             expected_data (list[dict]): Expected dataframe rows defined as dicts.
             fields (list[str]): Compare only certain fields.
             ordered (bool): Does order of rows matter?
+            places (int|None): Compare floats up to given decimal places.
```
I'd prefer to set this as allowed deviation (float), like abs(x'-x) < dev, this is more precise and common. The profit also that you can call the parameter with a greek letter 🇬🇷  😄  eps (epsilon) or delta

mantzouratos replied:
Hahaha! Using a greek letter was a strong argument, but I opted to follow NumPy's convention instead (rtol/atol as in https://docs.scipy.org/doc/numpy/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose)

skliarpawlo replied:
Shi why to follow such confusing api's :) I really don't like it much more than your impl ;) , they looks to mimic some C lib api like matlab which is awful. Also does parameter name 'places' forces to look into docs only me? Am I able to test that deviation is not more than 0.2 for example?

mantzouratos replied:
Hm when comparing floats you do need two switches: relative tolerance (deviation is not more than 0.2) and absolute tolerance (deviation is not more than 10% of float's magnitude). I think NumPy's API is pretty standard.

With our previous API, we used `places` and `significant` as proxies to achieve the same effect. Where you could round up to X decimal places or X significant digits. It was more confusing that the current API, so I got rid of it completely - we don't have `places` anymore (although python's `assertAlmostEqual` also uses it). 

As said, yes, now you can check deviation is not more than 0.2 by just setting `atol=0.2`.

skliarpawlo commented on sparkly/testing.py:
```diff
@@ -120,28 +147,132 @@ def tearDown(self):
+                ignore order arbitrarily deep.
+            places (int|None): Decimal or significant digits to consider
+                during comparisons of floats.
+            significant (bool|False): If set, interpret places as # of
```
quite complex api, key params are not quite self-documenting

mantzouratos replied:
changed it quite a bit, and added more explanations. hopefully it's cleaner now!

albertc1 commented on sparkly/testing.py:
```diff
@@ -120,28 +147,132 @@ def tearDown(self):
+        Extend this classic function signature to work better with
+        comparisons involving dataframes, rows, dictionaries and lists by
+        ignoring the order of mutable nested structures recursively and
+        rounding up floats when necessary.
```
you should probably also note that as a side effect, lists can sometimes assertEqual to dicts, and Datatypes and floats assertEqual to certain strings, etc.

mantzouratos replied:
Thanks for pointing this out - with the new `assertEqual` implementation this side effect won't happen.

albertc1 commented on sparkly/testing.py:
```diff
@@ -120,28 +147,132 @@ def tearDown(self):
+
+                obj = mock.call(
+                    *[_nicefy(call_arg, recursion_depth + 1) for call_arg in call_args],
+                    **collections.OrderedDict(_nicefy(call_kwargs, recursion_depth + 1))
```
why does this need to be an ordered dict?  the order of kwargs doesn't matter, right?
`mock.call(a, b=1, c=2)` is equal to `mock.call(a, c=2, b=1)`

mantzouratos replied:
Yeah, it was a hack to make `_create_key` work. Got rid of support for `mock.calls` since they're not related to Spark in any way, and it's easy to build it when needed on top of the new `assertEqual`

albertc1 commented on sparkly/testing.py:
```diff
@@ -120,28 +147,132 @@ def tearDown(self):
+            return entry is None, entry
+
+        def _nicefy(obj, recursion_depth=0):
+            if isinstance(obj, float):
```
This function is adding multiple separate improvements to assertEqual, and is a bit difficult to understand as a result.  I'd pull out the logic for floats and T.DataTypes into their own functions that you register using TestCase.addTypeEqualityFunc() (https://docs.python.org/2/library/unittest.html#unittest.TestCase.addTypeEqualityFunc).  The float precision params can be configured using an private instance variable. 

This way, your function is more focused around adding the ability to compare and ignore order in complex datastructures.

mantzouratos replied:
Thanks for pointing `addTypeEqualityFunc` out! Refactored the implementation and the API as a result and hopefully now it will be much easier to follow (and use).

mantzouratos commented on sparkly/writer.py:
```diff
@@ -255,6 +275,173 @@ def write_partition_to_kafka(messages):
+                same redis key together in an array before exporting. By
+                default if multiple rows share the same redis key, one
+                will overwrite the other.
+            expire (int|None): Expire the keys after this number of seconds.
```
I preferred `expire` to keep consistent with Redis jargon that refers to this as just `expire`. E.g., https://redis.io/commands/expire, https://github.com/andymccurdy/redis-py/blob/master/redis/client.py#L1142

albertc1 replied:
`expire` refers to the action you are doing, not the parameter of the action, which is the ttl.  You want to expire the key after TTL seconds.  There is also a TTL function (https://redis.io/commands/ttl) that fetches what this parameter represents.

mantzouratos replied:
The official redis client refers to this as `ex`(pire) in their `set(...)` function linked above. If you want to pass a timestamp there is an `ExpireAt` command in the official documentation, which would correspond to an `expireAt` URL parameter. I think this is a pretty straightforward association with Redis' official manuals, so I don't think `expire` is ambiguous at all. `TTL` provides read access to whatever the expire flag is and as such I think is more confusing to use in this case.

albertc1 commented on sparkly/testing.py:
```diff
@@ -214,65 +235,234 @@ def assertEqual(self,
+        temp_add_type_equality_func = partial(temp_add_type_equality_func, self)
+
+        # Register equality asserters
+        with temp_add_type_equality_func(float, partial(assert_float_equal, self)), \
```
What if instead of overriding the assertEqual function, we just register these type_equality_functions in `__init__()` and configure atol/rtol/ignore_nullability/ignore_order/ignore_order_depth via instance attributes, similar to how maxDiff, longMessage, and failureException are configured?

We'd need to write equality functions for list and dict primitives that just look like 
```
def assert_dict_equal(self, a, b, msg=None):
    self.assertEqual(Dict(a), Dict(b), msg)
```

We could add some simple helper functions (set_float_tolerance(), etc.) to make it easier to update those config values, but I think in most cases, those should be the same for almost all calls in a function, so it'd be easier to configure it once and then not worry about it. 
You also wouldn't need re/un-register these functions this way.

mantzouratos replied:
Not a fan of this idea, as I think it makes it more tedious for the end user and makes the final code and interface more cluttered. Parameters like float comparison tolerance have traditionally been part of `assertEqual...` statements (e.g., Python standard lib, NumPy testing) and it's counter-intuitive / uncommon to have them set as class attributes for which you need to call an extra function to set. Similar for the other parameters - I feel they belong to `assertEqual`.

Subsequently, the implementation of (some of) these classes and/or their asserters depend on these parameters and might only be used within the scope of one call of this function. I like the re/un-registering pattern because
- it adds negligible overhead
- it's simple to understand and a simple invariant to maintain
- it cleans up for the next equality assertion no matter what problems come up and no matter how the user wants to mix our `assertEqual` variation with the vanilla one

albertc1 replied:
i actually think it's more cumbersome to have to pass an "ignore_order" and "ignore_nullability" and float precision arguments to all of your assert_equals calls, instead of just setting it once like we do with maxDiff in a lot of places.  I agree that it would be cumbersome if you expect to call assertEqual with different values for those paramters pretty often in the same test, but I think that will be the exceptional case.  

I think those params make sense as part of an assertEqual call when the assert equal call is specific to a single type of input, like assertAlmostEqual() or assertListEqual(), but now you are adding 5 different params that govern 3 different types of functionality into one interface which makes it much more confusing.  The param names like atol and rtol are only NOT confusing in cases like numpy.testing because you are calling them in functions where the whole purpose is to do comparison with tolerances.  

If you really want to make it easy to mix your assertEqual with vanilla assertEqual you shouldn't be overriding it; just name this function something different.  Unregistering isn't really necessary because registering only registers for this instance of the unittest.testcase.  Other testcases can still use vanilla assertEqual.

mantzouratos replied:
In my opinion, the best practice is for each test to be as self-contained as possible and avoid any magic / hidden parameter settings so that the user can just read it and immediately understand _exactly_ the test conditions etc., so I would probably appreciate seeing `ignore_nullability=True` set explicitly every time you use this. However, if you want to set some arguments to other default values once and for all, you could still do it pretty easily in your `setUp`:
```
from functools import partial


class MyTestCase(SparklyTest):

    def setUp(self):
        self.assertEqual = partial(self.assertEqual, atol=15, ...)
```

I don't think atol and rtol are confusing if somebody is either familiar with numpy or it's the second time they use this function; sure, the first time it might be but what isn't. Dunno, I guess other opinions are welcome on this matter.

Unregistering is necessary because on one test case you might want to use a specific tolerance and on the other use another tolerance (or another `ignore_depth` level).

mantzouratos commented on sparkly/testing.py:
```diff
@@ -172,40 +176,57 @@ def assertDataFrameEqual(self,
+                ignore order infinitely, 1 means ignore order only at the
+                top level, 2 will ignore order within lists of lists and
+                so on. Default is ignore order arbitrarily deep.
+            atol (int, float|0): Absolute tolerance in float comparisons.
```
I prefer to use a shortcut that is standard when talking about tolerances (e.g. http://web.mit.edu/10.001/Web/Tips/Converge.htm) over the full long name that is cumbersome to write. Also I think an API with atol/rtol is less confusing than places/delta.

mantzouratos commented on sparkly/testing.py:
```diff
@@ -214,65 +235,234 @@ def assertEqual(self,
+            return value
+
+        atol = atol or 0
+        rtol = rtol or 0
```
Yeah but I'll just remove them. I guess I was being overly cautious.

mantzouratos commented on sparkly/testing.py:
```diff
@@ -214,65 +235,234 @@ def assertEqual(self,
+        rtol = rtol or 0
+
+        @total_ordering
+        class Float(object):
```
The definitions of these classes (especially Floats) might depend on the arguments specified in the `assertEqual` call. I think it's cleaner this way than defining them outside of the function since they are scoped inside the only namespace they are actually used in. Having the high level plan presented in the beginning of the function helps to keep focus while going through it.

mantzouratos commented on sparkly/testing.py:
```diff
@@ -214,65 +235,234 @@ def assertEqual(self,
+            if float1 == float2:
+                return
+
+            if not atol and not rtol:
```
Yes.

drudim commented on sparkly/writer.py:
```diff
@@ -255,6 +275,188 @@ def write_partition_to_kafka(messages):
+                return zlib.compress(data)
+            elif protocol == 'gzip':
+                try:
+                    return gzip.compress(data)
```
are you sure we don't need to call `flush` here? https://github.com/python/cpython/blob/master/Lib/gzip.py#L302

mantzouratos replied:
Yes. The library takes care of it for us. See https://github.com/python/cpython/blob/master/Lib/gzip.py#L523-L524. This part is already in production by the way

drudim commented on sparkly/writer.py:
```diff
@@ -255,6 +275,188 @@ def write_partition_to_kafka(messages):
+                    if group_by_key:
+                        row = row[value_name]
+                    if exclude_null_fields:
+                        for null_field in list(f for f in row.keys() if row[f] is None):
```
what is the reason to transform to list here?
`for null_field in (key for key in row.items() if value is None):` does the same but in more efficient way.

mantzouratos replied:
If you don't evaluate to a list you'll get a `RuntimeError: dictionary changed size during iteration` in python 3.

drudim commented on sparkly/testing.py:
```diff
@@ -135,13 +172,294 @@ def assertDataFrameEqual(self, actual_df, expected_data, fields=None, ordered=Fa
+            equal_nan=False,
+        )
+
+    def assertEqual(self,
```
What is the point to redefine builtin `assertEqual`? Isn't it better to emphasize that our implementation is different?

mantzouratos replied:
I think it's cleaner for the end code to only have to use one function for all their equality assertions. Easy to read, blends well with other `assertEqual` calls in the same test, easy to know what it is doing just by looking at the name.

drudim replied:
I understand your point, but, unfortunately, can't agree with it. Explicit is better than implicit.
We can vote for the final decision (@albertc1 @skliarpawlo ?)

mantzouratos replied:
Ok. I'd call for @srstrickland and @tingyu-h in the voting too since they've been doing a lot of spark work

srstrickland replied:
@drudim do you have a proposal for a better name?

I am used to using specialization functions like `assertDictEqual` and `assertListEqual`, primarily for their better output, so I would be OK to use another `assertFooEqual` here, but it's not limited to a single type so I'm not sure what the name should be?

drudim replied:
How about `assertDataFrameEqual`? There is no need to make it god-like function that manages everything. We have `assertAlmostEqual` to compare `float` and `assertCountEqual` to compare `list` without preserving the order.

albertc1 replied:
I vote for changing the name instead of overriding default method

mantzouratos replied:
I don't think `assertDataFrameEqual` is a good choice here since:
* this function is not limited to dataframes.
* this would introduce a breaking change in the API since we have already defined `assertDataFrameEqual`

@drudim, in your example, you still need to compose somehow `assertCountEqual` and `assertAlmostEqual` if you want to be able to compare lists of floats right? So I don't understand what you're trying to say with this comparison.

I think `assertEqual` is the best name because it makes for very clean, readable tests, very easy to plug in existing tests / projects without changing code, and just feels the right level of abstraction.

drudim replied:
> this would introduce a breaking change in the API since we have already defined `assertDataFrameEqual`

Why is this a breaking change?

> if you want to be able to compare lists of floats right? So I don't understand what you're trying to say with this comparison.

Why `sparkly` should take care about such comparators? The library is designed to work with `pyspark`. The main data structure is `DataFrame`. There are different `assert` functions to work with different data structures, we can't (and shouldn't) cover all the world.

mantzouratos replied:
`sparkly` takes care about such comparators because **they can and will** be part of data you collect from a `DataFrame`. It's very tedious to have to compose all these different `assert` functions together every time you want to compare rows you collected that contain arbitrarily nested arrays, maps and floats.

drudim commented on sparkly/testing.py:
```diff
@@ -135,13 +172,294 @@ def assertDataFrameEqual(self, actual_df, expected_data, fields=None, ordered=Fa
+        # for some, and then feed them to the vanilla assertEqual.
+
+        # Define our custom containers
+        def cast_to_test_friendly_container(value, ignore_order_depth):
```
did you consider to use http://pythonhosted.org/testfixtures/comparing.html#providing-your-own-comparers ? IMO, `textfixtures.compare` provides much more visibility on diff reports in case of complex structures.

drudim replied:
`compare` is very useful in case when you have an array of structs, and some value is different within a record.

mantzouratos replied:
How about this?
```
E   AssertionError: Lists differ: [Row([23 chars]hello', 'entity_type'='yo'), Row('entity_gid'=[26 chars]')])] != [Row([23 chars]hellop', 'entity_type'='yo'), Row('entity_gid'[27 chars]')])]
E   
E   First differing element 0:
E   Row('a'=[Row('entity_gid'='hello', 'entity_type'='yo'), Row('entity_gid'='hi', 'entity_type'='there')])
E   Row('a'=[Row('entity_gid'='hellop', 'entity_type'='yo'), Row('entity_gid'='hi', 'entity_type'='there')])
E   
E   - [{'a': [Row('entity_gid'='hello', 'entity_type'='yo'),
E   + [{'a': [Row('entity_gid'='hellop', 'entity_type'='yo'),
E   ?                                +
E   
E             Row('entity_gid'='hi', 'entity_type'='there')]}]
```

drudim replied:
that works for me

drudim commented on sparkly/utils.py:
```diff
@@ -59,6 +64,89 @@ def absolute_path(file_path, *rel_path):
+    """Check whether a complex dataType has specific fields.
+
+    Args:
+        t (pyspark.sql.types.ArrayType, pyspark.sql.types.MapType or
```
is it hard to support other primitive types?

mantzouratos replied:
It might not be hard, but what would be the usecase? I can't think of any.

drudim replied:
don't have a use-case, but the signature will be much simpler in this case:
```
t (DataType) instead of t(ArrayType, MapType, StructType)
```
I am ok to leave it as is.

drudim commented on sparkly/utils.py:
```diff
@@ -59,6 +64,89 @@ def absolute_path(file_path, *rel_path):
+        yield chunk
+
+
+def type_has(t, required_fields):
```
How about to make another util that accepts `DataFrame`, e.g.:
```
def dataframe_has(df, required_fields):
    return type_has(df.schema, required_fields)
```

mantzouratos replied:
Other than the fact that `dataframe_has` seems to be a better name it doesn't save us anything else (just the `.schema` part), so I'm not sure it's worth to have to maintain it (and write tests etc). Maybe rename `type_has` to `datatype_has`?

drudim replied:
how about `schema_has`? 

> Other than the fact that dataframe_has seems to be a better name it doesn't save us anything else (just the .schema part)

I am ok to keep it simple.

drudim commented on sparkly/utils.py:
```diff
@@ -59,6 +64,89 @@ def absolute_path(file_path, *rel_path):
+    Raises:
+        AssertionError, if t and required_fields cannot be compared
+           because they aren't instances of the same complex dataType.
+        KeyError, if a required field is not found in the struct.
```
I am not sure about using default exceptions here. Two problems:
- hard to understand when something goes wrong in the function itself
- hard to provide a good feedback on the error

mantzouratos replied:
I like using default exceptions because they have a non-ambiguous, pre-defined meaning, and the user won't need to import any special exceptions to be able to catch them. Also I think these ones lend themselves naturally to the errors they signal in this case. Even spark structures like `Row` use them.

To your two points:
* nothing should go wrong in the function itself, but if it does it will look like it in the traceback. But this should happen extremely rarely.
* Feedback on the error is very clear imo. For key errors it's the same you get when a key doesn't exist in a dictionary (which is a similar operation) with us providing the full path to the missing field, and for type errors we explicitly state what we were expecting and what we saw.

drudim replied:
> nothing should go wrong in the function itself, but if it does it will look like it in the traceback. But this should happen extremely rarely.

It sounds like: "we should write code without bugs".

I guess in this specific case it's relatively safe, because usually we use it in assertions like:
```
def func(df):
    assert schema_has(df.schema, ...)
```

BUT, I will be strongly against of using generic exceptions for APIs like:
```
try:
     some_func()
except KeyError:
     # kill people
else:
     # love people
```

drudim commented:
We also need to add documentation for all methods that we are going to expose with this PR. But let's clarify what will be included first.

mantzouratos commented on sparkly/catalog.py:
```diff
@@ -26,6 +27,62 @@ def __init__(self, spark):
         """
         self._spark = spark
 
+    def create_table(self, table_name, path=None, source=None, schema=None, **options):
```
Instead of implementing similar functionality in internal plugins, why not add the metadata copying here after this is merged?

Also I'm not sure why we should explain the difference with `saveAsTable`. Imo pyspark api documentation is enough since we mention explicitly that this extends `createTable/createExternalTable`.

drudim replied:
> Instead of implementing similar functionality in internal plugins, why not add the metadata copying here after this is merged?

Because the logic of replacement via tmp table is very specific. It's our trick to update tables, not a general solution. 

> Also I'm not sure why we should explain the difference with saveAsTable. Imo pyspark api documentation is enough since we mention explicitly that this extends createTable/createExternalTable.

It isn't mandatory.

mantzouratos commented on sparkly/catalog.py:
```diff
@@ -26,6 +27,62 @@ def __init__(self, spark):
+
+        try:
+            createTable = self._spark.catalog.createTable
+        except AttributeError:  # before Spark 2.2
```
Well this is a usual python pattern and I don't think it's less readable

drudim replied:
- it isn't safe, `AttributeError` could be raised at `self._spark` or `self._spark.catalog`
- pythonic way will be to check for `hasattr(self._spark.catalog, 'createTable')`

mantzouratos commented on sparkly/functions.py:
```diff
@@ -0,0 +1,139 @@
+        dfs[idx] = df
+
+    # Join the dataframes
+    joined_df = reduce(partial(type(dfs[0]).join, on=on, how=how), dfs)
```
Haven't seen such behavior and we've been using this extensively with left outers in creator loader. Remember I had encountered an issue with outer joins where `null` wasn't properly recognized but I don't remember the exact details now..

mantzouratos commented on sparkly/functions.py:
```diff
@@ -0,0 +1,139 @@
+    return joined_df
+
+
+def switch_case(switch, case=None, default=None, **additional_cases):
```
`case` is the only way to compare your `switch` column to non-string values (e.g., integers, boolean, etc), which can't be expanded to keyword arguments.

drudim replied:
Fair enough, how about: `def switch_case(switch, cases, default=None)` then?
To simplify api.

mantzouratos replied:
If you feel strongly about it I can drop `**additional_cases` but it has been pretty convenient to be able to say `switch_case('state', CA='California', ...)` in the string cases (that are very frequent) and not have to use the `**{...}` notation.

mantzouratos commented on sparkly/writer.py:
```diff
@@ -255,6 +275,188 @@ def write_partition_to_kafka(messages):
 
         rdd.mapPartitions(write_partition_to_kafka).count()
 
+    def redis(self,
```
Reader would not be hard to add, but doesn't seem like a generic usecase. On the other hand, it seems much more natural to write from Spark to Redis. I don't really understand what "very specific" means in this context.. I think it's as specific as exporting to Cassandra or ES.

drudim replied:
@mantzouratos Redis supports different data structures. You picked only one of them to solve your use-case, but it doesn't mean that everybody will use it like that.

In addition to that, very specific decisions are:
- `keys` generation
- compression
- grouping into arrays

mantzouratos commented on sparkly/writer.py:
```diff
@@ -255,6 +275,188 @@ def write_partition_to_kafka(messages):
+                must be provided. See below.
+            port (int|6379): Port redis is listening to.
+            db (int|0): Redis db to write to.
+            redis_client_init (callable|None): Bypass internal redis
```
Yes, we need it, because redis-py sets up a client internally that it returns through `from_url` when you use it to parse `redis` and `rediss` URLs. So we're using it when resolving the URL. In this particular case it's the only extra thing that would end up in `options`, so I feel it's cleaner to have it here explicitly.

drudim replied:
sorry, didn't get about parsing, can I ask you to share URL to that part in redis-py documentation?

mantzouratos replied:
https://redis-py.readthedocs.io/en/latest/#redis.StrictRedis.from_url

mantzouratos commented on sparkly/writer.py:
```diff
@@ -255,6 +275,188 @@ def write_partition_to_kafka(messages):
+                    F.struct(*[col for col in df.columns if col != key_name])
+                )
+                .groupBy(key_name)
+                .agg(F.expr('collect_list( {} )'.format(value_name)).alias(value_name))
```
Good catch, stayed as an `expr` because I was using brickhouse's `collect` for this in spark 1.6

mantzouratos commented:
@drudim Thanks for the thorough review! Addressed your comments, either with changes or with new comments 🎉 . Please take a look again.

To use your punch list - for unpunched tasks pls see comments:
- [x] `multijoin` minor code fixes
- [x] `switch_case` minor code fixes
- [x] `create_table` not sure if we need it
- [ ] `lru_cache` I am sure we don't need it: _I think we do need it_
- [x] `redis writer` minor code fixes, will be moved to a plugin in the future
- [ ] `assertEqual` IMO we should change name + use `textfixtures` to simplify code and improve functionality: _ I think we should keep the name, please provide usecase where external pkg improves functionality_
- [ ] `type_has` IMO we should add support for minor types and introduce `dataframe_has` function: _ don't see when we'd need to compare minor types, nor why we should maintain `dataframe_has`_

mantzouratos commented:
retest this please

mantzouratos commented:
retest this please

drudim commented:
1) `lru_cache` Can you explain the value of such decorator?

How is it different from?
```
from functools import lru_cache

@lru_cache
def my_func():
    return df.cache()
```
Here you say that all we save is `.schema` https://github.com/Tubular/sparkly/pull/45#discussion_r127834376, but in case of `lru_cache` you are trying to save `.cache`

2) `assertEqual` replied about use-case, let's vote for naming.

3) `type_has` let's rename it to `schema_has`, I am ok to keep the current implementation as is.

mantzouratos commented:
1. Not sure I understand your comment or what you quoted? `lru_cache` is useful when you want to persist/unpersist dataframes programmatically, without having to manage it explicitly through code. No need to pollute the actual transformation with random `.persist(StorageLevel.DISK_ONLY)` statements or anything of that sort, and in the future we could even extend it to persist to S3 or something.

2. added a comment on the naming, i really feel `assertEqual` is the most suitable name but waiting for consensus.

3. renamed.

drudim commented:
1. I missed the part about `unpersist` call. Now it makes sense to me.
2. Still don't like that we replace the default `assertEqual`. I am ok to keep godlike comparator, but it should be named properly.

drudim commented:
> they can and will be part of data you collect from a DataFrame
> 

Right, but they are a part of DataFrame. So, let’s call the method properly.
As for the name, how about `assertRowsEqual`?

We also need to add warning for `assertDataFrameEqual` to mark it as deprecated.

> On Aug 9, 2017, at 7:56 PM, Yannis Mantzouratos <notifications@github.com> wrote:
> 
> @mantzouratos commented on this pull request.
> 
> In sparkly/testing.py <https://github.com/Tubular/sparkly/pull/45#discussion_r132242851>:
> 
> > -            self.assertEqual(actual_data, expected_data)
> -        else:
> -            try:
> -                self.assertCountEqual(actual_data, expected_data)
> -            except AttributeError:
> -                self.assertItemsEqual(actual_data, expected_data)
> +        return self.assertEqual(
> +            actual_data,
> +            expected_data,
> +            ignore_order=not ordered,
> +            ignore_order_depth=1,
> +            rtol=0,
> +            equal_nan=False,
> +        )
> +
> +    def assertEqual(self,
> sparkly takes care about such comparators because they can and will be part of data you collect from a DataFrame. It's very tedious to have to compose all these different assert functions together every time you want to compare rows you collected that contain arbitrarily nested arrays, maps and floats.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Tubular/sparkly/pull/45#discussion_r132242851>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AB1kJkLQBOhZD5rZu_N1n7nPCT77fpyLks5sWeSygaJpZM4OE3jz>.
>

mantzouratos commented:
Ok, renamed to `assertRowsEqual` and added the warning. Pls let me know if everything seems alright now. Once you approve I'll start working on documentation, then ping you guys again for a final round of approvals.

skliarpawlo commented on docs/source/reader_and_writer.rst:
```diff
@@ -172,6 +172,39 @@ Basically, it's just a high level api on top of the native
         'rewriteBatchedStatements': 'true',  # improves write throughput dramatically
     })
 
+.. _redis:
```
Sorry, missed the decision here: do we decided to have writer, and not having reader? + from the api it looks like you dictate that value is compressed json and how key is formatted - delimiter-separated. That is probably good defaults but I'd prefer to have it more generic. For example like in kafka: provide key-serializer and value-serializer as params, having your defaults, this will also make implementing reader much easier

mantzouratos replied:
Yes, we've decided we don't care about a reader currently since main usage of Redis is as a cache.

I've started implementing your key/value serializers suggestion as it does provide a more powerful api at the end of the day, but I stopped and I'm not sure it's a good idea anymore:
* It complicates code quite a lot.
* As a user I'd expect that `write_ext.redis` allows me to set similar parameters to the ones I can set through `write_ext.by_url`. That's true for all other datastores we implement here except for Kafka, for which we don't provide URL support at all.
* It becomes tedious from a user perspective to set simple serializing parameters like `compression` and `key_prefix` as now I have to define a function to do it.
* It provides more fine grained control than we will probably ever need. If the need arises internally or a user of the package requests it I'm all in for re-visiting.
* I've actually spent way more time on this PR than it was initially allocated so I can't invest much more.

So, unless you feel strongly about this, I'd prefer to stash it for a v2.

mantzouratos commented on sparkly/testing.py:
```diff
@@ -96,9 +96,7 @@ class SparklyTest(TestCase):
-    def setUpClass(cls):
-        super(SparklyTest, cls).setUpClass()
-
+    def _init_session(cls):
```
Unfortunately it's needed, otherwise anything else we set up for `SparklyTest` has to be replicated for `SparklyGlobalTest`. Currently that includes fixtures and `test_target` importing.
